pandas/tests/groupby/test_frame_value_counts.py

```diff
@@ -0,0 +1,254 @@
+        ),
+    )
+    tm.assert_series_equal(result_frame, expected)
+
+    animals_df["key"] = 1
+
+    result_frame_groupby = animals_df.groupby("key").value_counts(
+        sort=sort, ascending=ascending, normalize=normalize
+    )
+    result_frame_groupby.reset_index(drop=True, level="key", inplace=True)
+    result_frame_groupby.name = None
+    tm.assert_series_equal(result_frame_groupby, expected)
+
+
+@pytest.fixture
+def names_with_nulls_df(nulls_fixture):
+    return pd.DataFrame(
+        {
+            "first_name": ["John", "Anne", "John", "Beth"],
+            "middle_name": ["Smith", nulls_fixture, nulls_fixture, "Louise"],
+        },
+    )
+
+
+@pytest.mark.parametrize(
+    "dropna, expected_data, expected_index",
+    [
+        (
+            True,
+            [1, 1],
+            pd.MultiIndex.from_arrays(
+                [("Beth", "John"), ("Louise", "Smith")],
+                names=["first_name", "middle_name"],
+            ),
+        ),
+        (
+            False,
+            [1, 1, 1, 1],
+            pd.MultiIndex(
+                levels=[
+                    pd.Index(["Anne", "Beth", "John"]),
+                    pd.Index(["Louise", "Smith", np.nan]),
+                ],
+                codes=[[0, 1, 2, 2], [2, 0, 1, 2]],
+                names=["first_name", "middle_name"],
+            ),
+        ),
+    ],
+)
+@pytest.mark.parametrize("normalize", [False, True])
+def test_data_frame_value_counts_dropna(
+    names_with_nulls_df, dropna, normalize, expected_data, expected_index
+):
+    # GH 41334
+    # 3-way compare with :meth:`~DataFrame.value_counts`
+    # Tests with nulls from frame/methods/test_value_counts.py
+    result_frame = names_with_nulls_df.value_counts(dropna=dropna, normalize=normalize)
+    expected = pd.Series(
+        data=expected_data,
+        index=expected_index,
+    )
+    if normalize:
+        expected /= float(len(expected_data))
+
+    tm.assert_series_equal(result_frame, expected)
+
+    names_with_nulls_df["key"] = 1
+    result_frame_groupby = names_with_nulls_df.groupby("key").value_counts(
+        dropna=dropna, normalize=normalize
+    )
+    result_frame_groupby.reset_index(drop=True, level="key", inplace=True)
+    result_frame_groupby.name = None
+
+    tm.assert_series_equal(result_frame_groupby, expected)
+
+
+@pytest.mark.parametrize("as_index", [False, True])
+@pytest.mark.parametrize("observed", [False, True])
+@pytest.mark.parametrize("normalize", [False, True])
+def test_categorical(education_df, as_index, observed, normalize):
+    # Test categorical data whether or not observed
+    gp = education_df.astype("category").groupby(
+        "country", as_index=as_index, observed=observed
+    )
+    result = gp.value_counts(normalize=normalize)
+
+    if observed:
+        gp = education_df.groupby("country", as_index=as_index)
+        expected = gp.value_counts(normalize=normalize)
+        if as_index:
```

Comment:
why are you using the numpy assert here?

we *always* want to use pandas comparisons as this is returning a frame right?

Reply:
Done